packages/core/saved-objects/core-saved-objects-api-server-internal/src/lib/apis/update.ts

```diff
@@ -118,63 +180,163 @@ export const performUpdate = async <T>(
+    );
+    const migratedUpdatedSavedObjectDoc = migrationHelper.migrateInputDocument({
+      ...migrated!,
+      id,
+      type,
+      // need to override the redacted NS values from the decrypted/migrated document
+      namespace: savedObjectNamespace,
+      namespaces: savedObjectNamespaces,
+      attributes: updatedAttributes,
+      updated_at: time,
+      ...(Array.isArray(references) && { references }),
+    });
+
+    const docToSend = serializer.savedObjectToRaw(
+      migratedUpdatedSavedObjectDoc as SavedObjectSanitizedDoc
+    );
+
+    // implement creating the call params
+    const indexRequestParams = {
+      id: docToSend._id,
       index: commonHelper.getIndexForType(type),
-      ...getExpectedVersionProperties(version),
       refresh,
-      retry_on_conflict: retryOnConflict,
-      body: {
-        doc,
-        ...(rawUpsert && { upsert: rawUpsert._source }),
-      },
-      _source_includes: ['namespace', 'namespaces', 'originId'],
+      body: docToSend._source,
+      // using version from the source doc if not provided as option to avoid erasing changes in case of concurrent calls
+      ...decodeRequestVersion(version || migrated!.version),
       require_alias: true,
-    })
-    .catch((err) => {
+    };
+
+    const {
+      body: indexDocResponseBody,
+      statusCode,
+      headers,
+    } = await client.index(indexRequestParams, { meta: true }).catch((err) => {
       if (SavedObjectsErrorHelpers.isEsUnavailableError(err)) {
         throw err;
       }
       if (SavedObjectsErrorHelpers.isNotFoundError(err)) {
         // see "404s from missing index" above
         throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
       }
+      if (SavedObjectsErrorHelpers.isConflictError(err)) {
+        // flag the error as being caused by an update conflict
+        err.retryableConflict = true;
+      }
       throw err;
     });
+    // throw if we can't verify a 404 response is from Elasticsearch
+    if (isNotFoundFromUnsupportedServer({ statusCode, headers })) {
+      throw SavedObjectsErrorHelpers.createGenericNotFoundEsUnavailableError(id, type);
+    }
+    // client.index doesn't return the indexed document.
+    // Rather than making another round trip to elasticsearch to fetch the doc, we use the SO we sent
+    // rawToSavedObject adds references as [] if undefined
+    const updatedSavedObject = serializer.rawToSavedObject<T>(
+      {
+        ...docToSend,
+        ...indexDocResponseBody,
+      },
+      { migrationVersionCompatibility }
+    );
 
-  const { originId } = body.get?._source ?? {};
-  let namespaces: string[] = [];
-  if (!registry.isNamespaceAgnostic(type)) {
-    namespaces = body.get?._source.namespaces ?? [
-      SavedObjectsUtils.namespaceIdToString(body.get?._source.namespace),
-    ];
-  }
+    // NEXT: assume the updatedSavedObject is the updated object we can use to construct the final response.
```

Comment:
We probably don't need all the old comments, only those that describe the new update workflow.